### PR TITLE
feat: Add OpenFeature.Extensions.Hosting package

### PR DESCRIPTION
**Pull request by austindrenski**
Closes: open-feature/dotnet-sdk#264

___

## Related
- open-feature/spec#81
  - open-feature/ofep#32
  - open-feature/spec#227

## TODO:

- [ ] moar tests
  - [ ] evaluation context
    - [ ] `IHostingEnvironment` sample
  - [ ] hooks
  - [ ] providers:
    - [ ] none
    - [ ] one
    - [ ] multiple
- [ ] xml docs
- [ ] readme docs

### CODE REVIEW

**Comment by austindrenski** [comment]
FYI couple of known test failures:
1. DCO is intentionally broken by a temp commit containing a squash of #187 and #188. Once those PRs are merged, I'll rebase and elide the temp commit.
2. dotnet-format is being weird about `OpenFeatureHostedService` and I'm pretty sure its just a sign that we need to migrate from the `dotnet-format` _tool_, to the new `dotnet format` that's baked into recent SDKs.
3. flaky unit tests: pretty sure these are all related to the async stuff, so hoping they'll be less common after #184 lands, but in the meantime please just re-run the test if its failed at the time of your review.

File: README.md [hunk 1]
```diff
@@ -67,6 +67,29 @@ public async Task Example()
 }
 ```
 
+### DI Usage
```
Comment (beeme1mr) [inline README.md:70]:
```suggestion
### Dependency Injection Usage
```
I suggest writing out the full name to avoid any confusion.

File: README.md [hunk 2]
```diff
@@ -67,6 +67,29 @@ public async Task Example()
...
+
+```csharp
+// Register your feature flag provider
+builder.Services.AddOpenFeature(static builder =>
+{
+    builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<FeatureProvider, SomeFeatureProvider>());
+    builder.TryAddOpenFeatureClient(SomeFeatureProvider.Name);
```
Comment (beeme1mr) [inline README.md:77]: Is it possible to do request scoped DI? It would be really great to include an example where you sent request specific context in the client. Since it's a slightly more advanced topic, perhaps it could be included in the evaluation context section later in the doc.

File: README.md
Comment (beeme1mr) [inline README.md:1]: We'll need to time this merge closely with the release or move this to a separate PR that can be merged after the 2.0 release.

File: README.md (hunk 2 above)
Comment (beeme1mr) [reply README.md:77]: It looks like it should work. Here's an example from .NET FeatureManagement.

https://github.com/microsoft/FeatureManagement-Dotnet?tab=readme-ov-file#using-httpcontext

File: src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+public sealed class OpenFeatureHostedService(Api api, IEnumerable<FeatureProvider> providers) : IHostedLifecycleService
+{
+    readonly Api _api = Check.NotNull(api);
+    readonly IEnumerable<FeatureProvider> _providers = Check.NotNull(providers);
+
+    async Task IHostedLifecycleService.StartingAsync(CancellationToken cancellationToken)
+    {
+        foreach (var provider in this._providers)
+        {
+            await this._api.SetProviderAsync(provider.GetMetadata().Name ?? string.Empty, provider).ConfigureAwait(false);
```
Comment (kinyoklion) [inline src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs:20]: I don't think the meta-data name for a provider is an appropriate name here. We probably need a way to make providers  that encapsulates that data.

Simple example is using one provider with two different environments.

But more than that the provider name is a different domain of names. It is the domain of the provider implementation, where client names are in the domain of the application. An application developer needs to know and assign that name.

Comment (toddbaert) [reply src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs:20]: :100:

File: src/OpenFeature.Extensions.Hosting/OpenFeatureBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+using Microsoft.Extensions.DependencyInjection;
+
+namespace OpenFeature;
+
+/// <summary>
+/// Describes a <see cref="OpenFeatureBuilder"/> backed by an <see cref="IServiceCollection"/>.
+/// </summary>
+/// <param name="ServiceCollection"><see cref="IServiceCollection"/></param>
+public sealed record OpenFeatureBuilder(IServiceCollection ServiceCollection);
```
Comment (arttonoyan) [inline src/OpenFeature.Extensions.Hosting/OpenFeatureBuilder.cs:9]: I would like to suggest renaming the `ServiceCollection` parameter to `Services`. The reasoning for this suggestion is to align with established .NET conventions where simpler, more generic names are often preferred for widely applicable interfaces or types. This naming convention is consistent with other .NET foundational libraries.

For instance: 

- in Microsoft.Extensions.Logging, the builder uses ILoggingBuilder with methods that refer to services in a general sense ([see the implementation here](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/ILoggingBuilder.cs)). 
- OpenTelemetry uses IOpenTelemetryBuilder, emphasizing a streamlined naming convention ([see implementation here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api.ProviderBuilderExtensions/IOpenTelemetryBuilder.cs#L12)).

File: src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace OpenFeature.Internal;
+
+internal sealed class OpenFeatureHostedService(Api api, IEnumerable<FeatureProvider> providers) : IHostedLifecycleService
+{
+    readonly Api _api = Check.NotNull(api);
+    readonly IEnumerable<FeatureProvider> _providers = Check.NotNull(providers);
+
+    async Task IHostedLifecycleService.StartingAsync(CancellationToken cancellationToken)
+    {
```
Comment (arttonoyan) [inline src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs:14]: I recommend adding a check for the CancellationToken at strategic points within our loops or before initiating long-running tasks. This would enhance our ability to respond to cancellation requests more gracefully.

```
// Check if cancellation has been requested before proceeding
cancellationToken.ThrowIfCancellationRequested();

```
Also, it might be beneficial to create an issue in OpenFeature to discuss the incorporation of CancellationToken parameters across all asynchronous methods. This would align our practices with best practices for asynchronous .NET APIs and improve the overall reliability and user control of the system.

File: src/OpenFeature.Extensions.Hosting/OpenFeatureBuilder.cs (hunk 4 above)
Comment (askpt) [reply src/OpenFeature.Extensions.Hosting/OpenFeatureBuilder.cs]: Renamed here: [Renamed per suggestion.](https://github.com/open-feature/dotnet-sdk/pull/181/commits/cf1f05fd6d0c3464b9c1a9c2e7eda1b4a658a90b)

File: src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs (hunk 5 above)
Comment (askpt) [reply src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs:14]: Added cancellation token: https://github.com/open-feature/dotnet-sdk/pull/181/commits/6bd175b58f1b2603165a34c8bfdb68ba93f54525

**Comment by nbr28** [comment]
Is this PR going to be merged? I was looking forward to this capabilities with version 2.0

**Comment by toddbaert** [comment]
> Is this PR going to be merged? I was looking forward to this capabilities with version 2.0

@nbr28 This is still being worked on, mostly by @askpt :pray: . We are still planning to merge once it's completed.

File: src/OpenFeature.Extensions.Hosting/OpenFeature.Extensions.Hosting.csproj [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <Nullable>enable</Nullable>
+    <TargetFrameworks>netstandard2.0;net6.0;net7.0;net8.0;net462</TargetFrameworks>
```
Comment (kylejuliandev) [inline src/OpenFeature.Extensions.Hosting/OpenFeature.Extensions.Hosting.csproj:5]: Should .NET 7 be included as TFM? [.NET 7 is out of official support](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core)

File: src/OpenFeature.Extensions.Hosting/OpenFeatureBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+using Microsoft.Extensions.DependencyInjection;
+
+namespace OpenFeature;
+
+/// <summary>
+/// Describes a <see cref="OpenFeatureBuilder"/> backed by an <see cref="IServiceCollection"/>.
+/// </summary>
+/// <param name="Services"><see cref="IServiceCollection"/></param>
+public sealed record OpenFeatureBuilder(IServiceCollection Services);
```
Comment (lukas-reining) [inline src/OpenFeature.Extensions.Hosting/OpenFeatureBuilder.cs:9]: What is the reason to go for a record and the extension methods here? I am not too deep into .NET but I have not seen that pattern formerly.

**Comment by askpt** [comment]
Superseded by #310
